### PR TITLE
NIFI-14005 - Support Custom Content from External File in GenerateFlowFile

### PR DESCRIPTION
**Pull request by pvillard31**
# Summary

[NIFI-14005](https://issues.apache.org/jira/browse/NIFI-14005) - Support Custom Content from External File in GenerateFlowFile

Add a property in GenerateFlowFile allowing the user to specify a local path to an external file. If specified the content of the external file will be used as content of the generated flow files.

This makes things easier compared to a GetFile for example. Besides, coupled with the Asset feature, it makes it easy to load a test dataset as content for GenerateFlowFile processor in a cloud native environment (container based).

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

**Review by exceptionfactory — COMMENTED** [review]
Thanks for putting together this improvement @pvillard31.

This change has implications for the permissions required to use the `GenerateFlowFile` Processor, as other Processors with access to the file system require the `READ_FILESYSTEM` permission. Although it is easy to add, it may have implications for existing flow designs. What do you think?

File: nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/GenerateFlowFile.java [hunk 1]
```diff
@@ -183,12 +200,22 @@ protected Collection<ValidationResult> customValidate(ValidationContext validati
...
 
         if (isCustom && (isUnique || !isText)) {
             results.add(new ValidationResult.Builder().subject("Custom Text").valid(false).explanation("If Custom Text is set, then Data Format must be "
                     + "text and Unique FlowFiles must be false.").build());
         }
+        if (isCustomContent && (isUnique || !isExternal)) {
+            results.add(new ValidationResult.Builder().subject("Custom Content").valid(false).explanation("If Custom Content is set, then Data Format must be "
+                    + "External File and Unique FlowFiles must be false.").build());
+        }
+        if (isCustomContent && isCustom) {
```
Comment (EndzeitBegins) [inline nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/GenerateFlowFile.java:215]: Do we need this validation explicitly? The `dependsOn` conditionals should result in only on of `Custom Text` or `Custom Content` being available to the user I assume?

**Comment by pvillard31** [comment]
That is an extremely valid point that I completely missed. Given that some users definitely want to restrict processors with access to local file system, it feels like we shouldn't change expectations around GenerateFlowFile and a dedicated processor for this specific usage would be better. What do you think?

File: nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/GenerateFlowFile.java [hunk 2]
```diff
@@ -111,8 +116,19 @@ public class GenerateFlowFile extends AbstractProcessor {
...
             .expressionLanguageSupported(ExpressionLanguageScope.ENVIRONMENT)
+            .dependsOn(DATA_FORMAT, DATA_FORMAT_TEXT)
+            .dependsOn(UNIQUE_FLOWFILES, "false")
             .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
             .build();
+    public static final PropertyDescriptor CUSTOM_CONTENT = new PropertyDescriptor.Builder()
+            .name("Custom Content")
+            .description("Path to a file. If specified, this file will be used as content of the generated FlowFiles and the File Size will be ignored.")
+            .required(false)
+            .identifiesExternalResource(ResourceCardinality.SINGLE, ResourceType.FILE, ResourceType.URL, ResourceType.TEXT)
```
Comment (EndzeitBegins) [inline nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/GenerateFlowFile.java:127]: Doesn't this warrant declaring the processor to require `READ_FILESYSTEM` permission? Is that a problem in regards to backwards compatibility?

**Review by EndzeitBegins — COMMENTED** [review]
Thank you for working on this Pierre.

This improvement will help with both manual and automated testing and allow to inject larger and non-text payloads more easily.

The overall changes look good to me. I've just left some remarks in regards to naming and declaration.

File: nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/GenerateFlowFile.java (hunk 2 above)
Comment (EndzeitBegins) [reply nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/GenerateFlowFile.java:127]: Looks like @exceptionfactory beat me to it.

**Comment by exceptionfactory** [comment]
> That is an extremely valid point that I completely missed. Given that some users definitely want to restrict processors with access to local file system, it feels like we shouldn't change expectations around GenerateFlowFile and a dedicated processor for this specific usage would be better. What do you think?

Yes, under the circumstances, it seems like a separate Processor would be better. I think it would be good to scope out what it looks like on the Jira issue and then move to consider the implementation.

**Comment by EndzeitBegins** [comment]
> a dedicated processor for this specific usage would be better. What do you think?

Thinking about it, a separate processor might make sense insofar that the processor is not actually "generating" content but rather injecting pre-defined content when using the contents of a file (although the same could be said about the existing "Custom Text")

**Pull request closed without merging**